round-based/src/state_machine/mod.rs

```diff
@@ -4,6 +4,33 @@
 //! may not be possible/desirable to have async runtime which drives the futures until completion.
 //! For such use-cases, we provide [`wrap_protocol`] function that wraps an MPC protocol defined as
 //! async function and returns the [`StateMachine`] that exposes sync API to carry out the protocol.
+//!
+//! ## Example
+//! ```rust,no_run
+//! use round_based::{Mpc, PartyIndex};
+//!
+//! # type Result<T, E = ()> = std::result::Result<T, E>;
+//! # type Randomness = [u8; 32];
+//! # type Msg = ();
+//! // Any MPC protocol
+//! pub async fn protocol_of_random_generation<M>(
+//!     party: M,
+//!     i: PartyIndex,
+//!     n: u16
+//! ) -> Result<Randomness>
+//! where
+//!     M: Mpc<ProtocolMessage = Msg>
+//! {
+//!     // ...
+//! # todo!()
+//! }
+//!
+//! let state_machine = round_based::state_machine::wrap_protocol(
+//!     |party| protocol_of_random_generation(party, 0, 3)
+//! );
+//! // `state_machine` implements `round_based::state_machine::StateMachine` trait.
+//! // Its methods can be used to advance protocol until completion.
```

Comment:
Maybe add an example of polling to completion? Maybe it should be in StateMachine itself, and here we can write "see example there"

Reply:
That's a bit difficult to do, as what "advancing until completion" looks like depends on how it's used, e.g. how incoming/outgoing messages are handled, etc. I could write something like:

```rust
let mut state = state_machine;

loop {
    match state.proceed() {
        ProceedResult:: SendMsg(msg) => outgoings.send(msg).await?,
        ProceedResult::NeedsOneMoreMessage => {
            let msg = incomings.next().await?;
            state.received_msg(msg)?;
        }
        ProceedResult::Yielded => {},
        ProceedResult::Output(out) => break Ok(out),
        ProceedResult::Error(err) => break Err(err),
    }
}
```

but that's a bit weird, as using async stream/sink defeats the purpose of the state machines

Reply:
On other hand, it might be a good thing to have this example anyway, to know what to do with the state machine in general...

Reply:
Exactly my thoughts, state machine polling is a difficult api and right now completely lacks examples

Reply:
Yeah true, I'll add this example

Reply:
Added an example. Lmk if you think it can be improved @maurges